src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationManager.cs

```diff
@@ -259,7 +234,78 @@ IEnumerator IEnumerable.GetEnumerator()
             }
         }
 
-        private class ConfigurationBuilderProperties : IDictionary<string, object>
+        private sealed class ProviderManager : IDisposable
+        {
+            private readonly object _replaceProvidersLock = new object();
+            private ProvidersReference _refCountedProviders = new(new List<IConfigurationProvider>());
+
+            public IEnumerable<IConfigurationProvider> Providers => _refCountedProviders.Providers;
+
+            public ProvidersReference GetReference()
+            {
+                // Lock to ensure oldRefCountedProviders.Dispose() in ReplaceProviders() doesn't decrement ref count to zero
+                // before calling _refCountedProviders.AddRef().
+                lock (_replaceProvidersLock)
+                {
+                    _refCountedProviders.AddRef();
+                    return _refCountedProviders;
+                }
+            }
+
+            // Providers should never be concurrently modified. Reading during modification is allowed.
+            public void ReplaceProviders(List<IConfigurationProvider> providers)
+            {
+                ProvidersReference oldRefCountedProviders = _refCountedProviders;
+
+                lock (_replaceProvidersLock)
+                {
+                    _refCountedProviders = new ProvidersReference(providers);
+                }
+
+                oldRefCountedProviders.Dispose();
+            }
+
+            public void AddProvider(IConfigurationProvider provider)
+            {
+                // Maintain existing references, but replace list with copy containing new item.
+                _refCountedProviders.Providers = new List<IConfigurationProvider>(_refCountedProviders.Providers)
+                {
+                    provider
+                };
+            }
+
+            public void Dispose() => _refCountedProviders.Dispose();
+        }
+
+        private sealed class ProvidersReference : IDisposable
+        {
+            private long _refCount = 1;
+
+            public ProvidersReference(List<IConfigurationProvider> providers)
+            {
+                Providers = providers;
+            }
+
+            public List<IConfigurationProvider> Providers { get; set; }
+
+            public void AddRef()
+            {
+                Interlocked.Increment(ref _refCount);
```

Comment:
What happens if _refCount was already 0 so this was already disposed?  Is that just something we don't care about? Or it's not possible by construction of consumers?

Reply:
`AddRef()` is only called by `ProviderManager.GetReference()` with the `_replaceProvidersLock` lock to ensure that this doesn't happen.

```
                // Lock to ensure oldRefCountedProviders.Dispose() in ReplaceProviders() doesn't decrement ref count to zero
                // before calling _refCountedProviders.AddRef().
```

I'll add a comment here too.

It should not be possible for the `_refCount` to go to zero until `ReplaceProviders` calls `oldRefCountedProviders.Dispose()` or the entire `ConfigurationManager` is disposed.